fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/JavaUserDefinedFunction.java

```diff
@@ -0,0 +1,127 @@
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.cascades.values;
+
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.query.plan.cascades.BuiltInFunction;
+import com.apple.foundationdb.record.query.plan.cascades.PromoteValue;
+import com.apple.foundationdb.record.query.plan.cascades.SemanticException;
+import com.apple.foundationdb.record.query.plan.cascades.typing.Type;
+import com.apple.foundationdb.record.query.plan.cascades.typing.Typed;
+import com.google.auto.service.AutoService;
+import com.google.common.base.Verify;
+import com.google.common.collect.ImmutableList;
+
+import javax.annotation.Nonnull;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Represents a Java user-defined function.
+ */
+@AutoService(BuiltInFunction.class)
+public class JavaUserDefinedFunction extends BuiltInFunction<Value> {
+    public JavaUserDefinedFunction() {
+        super("java_call", List.of(Type.primitiveType(Type.TypeCode.STRING)), new Type.Any(), JavaUserDefinedFunction::findFunction);
+    }
+
+    @SuppressWarnings({"unchecked", "DataFlowIssue"})
+    @Nonnull
+    private static Value findFunction(@Nonnull final BuiltInFunction<Value> builtInFunction, final List<? extends Typed> arguments) {
+        Verify.verify(!arguments.isEmpty());
+        Verify.verify(arguments.get(0).getResultType().getTypeCode().equals(Type.TypeCode.STRING));
+        // dispatching happens at query-building time, therefore, the argument must be literal
+        Verify.verify(arguments.get(0) instanceof LiteralValue<?>);
+        final var literalValue = (LiteralValue<?>)arguments.get(0);
+        final var functionName = (String)literalValue.compileTimeEval(null);
+
+        // for now, the function name is expected to represent the fully-qualified class name, so we can find
+        // it quickly via reflection, in the future we'll use the service loader to register the function.
+        Class<?> clazz;
+        try {
+            clazz = Class.forName(functionName);
+        } catch (ClassNotFoundException e) {
+            throw new RecordCoreException(String.format("could not find function '%s'", functionName), e);
+        }
+        // sanity / security check
+        if (!JavaUdf.class.isAssignableFrom(clazz)) {
+            throw new RecordCoreException(String.format("expecting '%s' to be a subclass of '%s'", clazz.getName(), JavaUdf.class.getName()));
+        }
+
+        final var fClazz = Verify.verifyNotNull(clazz);
+        // @param arguments The types of the arguments, must be the same type or of a promotable type of the corresponding subclass expected {@link Value} parameter(s).
+        Method method;
+        try {
+            method = fClazz.getMethod("getUdfParameterTypes");
```

Comment:
Why resolve a method using duck typing here on `JavaUdf`? You should resolve the user-supplied descendent of `BuiltinFunction` here, instantiate it or get it from a cache and then call its `encapsulate` method (or a derivate thereof).

Reply:
> Why resolve a method using duck typing here on JavaUdf?

It is not duck-typing, the provided type must be a descendant of `JavaUdf` as per line 68.